feat(stdlib): add system utility functions (uname, wc, sort, uniq, mount, pgrep)

rahuld109 opened the pull request:
Hey! I noticed issue #951 had been open for a while and figured I'd take a crack at it. Added a bunch of stdlib functions that were requested:

**What's included:**

- `uname_*` functions (kernel name, nodename, release, version, machine, os, all) - basically wrapping the uname command
- `lines_count`, `words_count`, `chars_count` - text counting utilities using wc
- `sort_lines`, `sort_lines_desc`, `sort_lines_numeric` - sorting helpers
- `uniq_lines`, `uniq_all_lines` - deduplication (adjacent vs all duplicates)
- `mount`, `umount`, `umount_force` - filesystem mounting (with `sudo` modifier)
- `pgrep`, `pgrep_exact`, `pkill`, `pkill_exact`, `pkill_force` - process management

Added tests for everything too. The pgrep tests are in the no_output directory since the results depend on what's running.

Closes #951
Partial fix for #954

rahuld109 committed:
feat(stdlib): add uname, wc, sort, uniq, mount/umount, pgrep/pkill functions

Added new stdlib functions:

env.ab:
- uname_kernel_name() - returns kernel name (e.g., Linux, Darwin)
- uname_nodename() - returns network hostname
- uname_kernel_release() - returns kernel release version
- uname_kernel_version() - returns kernel version
- uname_machine() - returns machine architecture
- uname_os() - returns OS name
- uname_all() - returns all uname info
- mount() - mounts a filesystem (requires root)
- umount() - unmounts a filesystem (requires root)
- umount_force() - force unmounts a filesystem
- pgrep() - finds process IDs by name pattern
- pgrep_exact() - finds process IDs by exact name
- pkill() - kills processes by name pattern
- pkill_exact() - kills processes by exact name
- pkill_force() - forcefully kills processes (SIGKILL)

text.ab:
- lines_count() - counts lines in text
- words_count() - counts words in text
- chars_count() - counts characters in text
- sort_lines() - sorts lines alphabetically
- sort_lines_desc() - sorts lines in descending order
- sort_lines_numeric() - sorts lines numerically
- uniq_lines() - removes consecutive duplicate lines
- uniq_all_lines() - removes all duplicate lines

Closes #951

rahuld109 committed:
fix: use const instead of let for immutable variables

rahuld109 committed:
fix(pgrep): properly read PIDs into array using while loop

rahuld109 committed:
test(pgrep): use 'sh' pattern for cross-shell compatibility

Tirito6626 commented:
pretty good, but usage examples should be updated (see https://github.com/amber-lang/amber/issues/954)

Tirito6626 commented:
you could also make use of [sudo modifier](https://docs.amber-lang.com/basic_syntax/commands#sudo) for functions that require root permsisions

Mte90 commented:
Cool!

The commands that require sudo are mount/umount. I will review the code asap!

rahuld109 committed:
fix: update usage examples to v0.6.0 syntax and add sudo modifier

- Changed echo to echo() in usage examples (v0.6.0+ syntax)
- Added sudo modifier to mount/umount functions for privilege escalation

rahuld109 commented:
Thanks for the feedback! I've updated the PR:

- Changed all `echo` to `echo()` in usage examples (v0.6.0+ syntax)
- Added `sudo` modifier to `mount()`, `umount()`, and `umount_force()` for intelligent privilege escalation

Tirito6626 commented:
code seems fine but i will also test it locally later

Mte90 commented on src/std/env.ab:
```diff
@@ -302,3 +302,173 @@ pub fun bash_version(): [Int] {
+#[allow_absurd_cast]
+pub fun pgrep(pattern: Text): [Int] {
+    let result = [Int]
+    trust $ while read -r pid; do {nameof result}+=("\$pid"); done < <(pgrep "{pattern}" 2>/dev/null) $
```
This while needs to be in amber code

rahuld109 replied:
Done - rewrote it using split_lines() and parse_int() instead

Mte90 commented on src/std/env.ab:
````diff
@@ -302,3 +302,173 @@ pub fun bash_version(): [Int] {
+/// pkill("nginx")?
+/// ```
+pub fun pkill(pattern: Text): Null? {
+    $ pkill "{pattern}" $?
````
return pkill output can be helpful?

Tirito6626 replied:
i think it would be also useful to be able to use PIDs and signal to send (e.g. SIGTERM or SIGINT, etc). `kill` could be used as fallback, since it's a bash builtin

rahuld109 replied:
Kept it as Null? for now since pkill doesn't output much useful info. Let me know if you'd prefer something different

rahuld109 replied:
Good idea - added a kill(pid, signal) function that defaults to TERM. Works with signal names or numbers

Mte90 reviewed (changes requested):
some changes needed but tests works with no errors

Tirito6626 committed:
Merge branch 'staging' into feat/stdlib-functions

rahuld109 committed:
refactor(stdlib): address review feedback for process functions

- Convert bash while loops to native Amber code in pgrep/pgrep_exact
- Add kill() function with PID and signal support (SIGTERM, SIGKILL, etc.)
- Use split_lines() and parse_int() for cleaner implementation

rahuld109 commented:
Hey, thanks for the feedback! I've made the changes:

- Rewrote the bash while loops in `pgrep`/`pgrep_exact` to use native Amber - now using `split_lines()` and `parse_int()` instead

- Added a `kill()` function that takes a PID and optional signal (defaults to TERM). Should cover the use cases you mentioned:
  ```ab
  kill(1234)?              // SIGTERM
  kill(1234, "SIGKILL")?   // or "9" works too
  ```

Let me know if there's anything else!

Tirito6626 commented:
why always using `nameof` instead of parsing the value directly?

rahuld109 commented:
This is already fixed in the latest commit - I removed the nameof/while loop and now it uses native Amber:

```amber
let output = trust $ pgrep "{pattern}" 2>/dev/null $
let lines = split_lines(output)
for line in lines {
    let pid = parse_int(line) failed { continue }
    result += [pid]
}
```

You might be seeing a cached diff - try refreshing the page

Tirito6626 commented:
i would also consider using here-string instead of printf pipe where possible, but i could do that after the merge, since i'd have to rework files anyways

Tirito6626 commented on src/std/text.ab:
````diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+/// ```
+#[allow_absurd_cast]
+pub fun chars_count(text: Text): Int {
+    return trust $ printf "%s" "\${nameof text}" | wc -m | tr -d ' ' $ as Int
````
here for example

Tirito6626 replied:
why using nameref here?

Tirito6626 commented on src/std/text.ab:
```diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+#[allow_absurd_cast]
+pub fun lines_count(text: Text): Int {
+    if text == "": return 0
+    return trust $ printf "%s\n" "\${nameof text}" | wc -l | tr -d ' ' $ as Int
```
```suggestion
    return trust $ wc -l <<< "{text}" $ as Int
```

Tirito6626 commented on src/std/text.ab:
````diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+/// ```
+#[allow_absurd_cast]
+pub fun words_count(text: Text): Int {
+    return trust $ printf "%s" "\${nameof text}" | wc -w | tr -d ' ' $ as Int
````
```suggestion
    return trust $ wc -w <<< "{text}" $ as Int
```

Tirito6626 commented on src/std/text.ab:
````diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+/// ```
+#[allow_absurd_cast]
+pub fun chars_count(text: Text): Int {
+    return trust $ printf "%s" "\${nameof text}" | wc -m | tr -d ' ' $ as Int
````
```suggestion
    return trust $ wc -m <<< "{text}" $ as Int
```

Tirito6626 commented on src/std/text.ab:
````diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+/// echo(sorted) // "apple\nbanana\ncherry"
+/// ```
+pub fun sort_lines(text: Text): Text {
+    return trust $ printf "%s" "\${nameof text}" | sort $
````
```suggestion
    return trust $ sort <<< "{text}" $
```

Tirito6626 commented on src/std/text.ab:
````diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+/// echo(sorted) // "cherry\nbanana\napple"
+/// ```
+pub fun sort_lines_desc(text: Text): Text {
+    return trust $ printf "%s" "\${nameof text}" | sort -r $
````
```suggestion
    return trust $ sort -r <<< "{text}" $
```

Tirito6626 commented on src/std/text.ab:
````diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+/// echo(sorted) // "1\n2\n10"
+/// ```
+pub fun sort_lines_numeric(text: Text): Text {
+    return trust $ printf "%s" "\${nameof text}" | sort -n $
````
```suggestion
    return trust $  sort -n <<< "{text}" $
```

Tirito6626 commented on src/std/text.ab:
````diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+/// echo(result) // "foo\nbar\nbaz"
+/// ```
+pub fun uniq_lines(text: Text): Text {
+    return trust $ printf "%s" "\${nameof text}" | uniq $
````
```suggestion
    return trust $ uniq <<< "{text}" $
```

Tirito6626 commented on src/std/text.ab:
````diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+/// echo(result) // "foo\nbar\nbaz"
+/// ```
+pub fun uniq_all_lines(text: Text): Text {
+    return trust $ printf "%s" "\${nameof text}" | awk '!seen[\$0]++' $
````
```suggestion
    return trust $ awk '!seen[\$0]++' <<< "{text}" $
```

rahuld109 committed:
refactor(stdlib): use here-strings instead of printf pipes

- Simplified sort_lines, sort_lines_desc, sort_lines_numeric
- Simplified uniq_lines, uniq_all_lines
- Simplified lines_count, words_count
- Keep printf for chars_count (here-string adds newline)

rahuld109 commented:
Applied the here-string suggestions - thanks for the cleaner approach!

- `sort_lines`, `sort_lines_desc`, `sort_lines_numeric` ✓
- `uniq_lines`, `uniq_all_lines` ✓
- `lines_count`, `words_count` ✓

Kept `chars_count` using printf since here-string adds a trailing newline which breaks the character count.

Tirito6626 reviewed (approved):
looks way better now, time to run tests

Ph0enixKM commented on src/std/env.ab:
````diff
@@ -369,3 +369,197 @@ pub fun bash_version(): [Int] {
+///
+/// ### Usage
+/// ```ab
+/// mount("/dev/sda1", "/mnt/disk")?
````
Would it be possible to add usage example for mount with options?

Ph0enixKM commented on src/std/text.ab:
````diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+
+/// Counts the number of lines in the given text.
+///
+/// ### Usage
+/// ```ab
+/// let count = lines_count("one\ntwo\nthree")
+/// echo(count) // 3
+/// ```
+#[allow_absurd_cast]
+pub fun lines_count(text: Text): Int {
+    if text == "": return 0
+    return trust $ wc -l <<< "{text}" | tr -d ' ' $ as Int
+}
````
I think that `count_lines` sounds better

```suggestion
/// ### Usage
/// ```ab
/// let count = count_lines("one\ntwo\nthree")
/// echo(count) // 3
/// ```
#[allow_absurd_cast]
pub fun count_lines(text: Text): Int {
    if text == "": return 0
    return trust $ wc -l <<< "{text}" | tr -d ' ' $ as Int
}
```

Ph0enixKM commented on src/std/text.ab:
````diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+/// Counts the number of words in the given text.
+///
+/// ### Usage
+/// ```ab
+/// let count = words_count("hello world foo")
+/// echo(count) // 3
+/// ```
+#[allow_absurd_cast]
+pub fun words_count(text: Text): Int {
+    return trust $ wc -w <<< "{text}" | tr -d ' ' $ as Int
+}
````
I think that `count_words` sounds better

```suggestion
/// ```ab
/// let count = count_words("hello world foo")
/// echo(count) // 3
/// ```
#[allow_absurd_cast]
pub fun count_words(text: Text): Int {
    return trust $ wc -w <<< "{text}" | tr -d ' ' $ as Int
}
```

Ph0enixKM commented on src/std/text.ab:
````diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+    return trust $ wc -w <<< "{text}" | tr -d ' ' $ as Int
+}
+
+/// Counts the number of characters in the given text.
+///
+/// ### Usage
+/// ```ab
+/// let count = chars_count("hello")
+/// echo(count) // 5
+/// ```
+#[allow_absurd_cast]
+pub fun chars_count(text: Text): Int {
+    return trust $ printf "%s" "{text}" | wc -m | tr -d ' ' $ as Int
+}
````
I feel like `count_chars` sounds better

```suggestion
/// Counts the number of characters in the given text.
///
/// ### Usage
/// ```ab
/// let count = count_chars("hello")
/// echo(count) // 5
/// ```
#[allow_absurd_cast]
pub fun count_chars(text: Text): Int {
    return trust $ printf "%s" "{text}" | wc -m | tr -d ' ' $ as Int
}
```

Ph0enixKM commented on src/std/text.ab:
````diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+    return trust $ printf "%s" "{text}" | wc -m | tr -d ' ' $ as Int
+}
+
+/// Sorts lines of text in ascending order.
+///
+/// ### Usage
+/// ```ab
+/// let sorted = sort_lines("banana\napple\ncherry")
+/// echo(sorted) // "apple\nbanana\ncherry"
+/// ```
+pub fun sort_lines(text: Text): Text {
+    return trust $ sort <<< "{text}" $
+}
+
+/// Sorts lines of text in descending order.
+///
+/// ### Usage
+/// ```ab
+/// let sorted = sort_lines_desc("banana\napple\ncherry")
+/// echo(sorted) // "cherry\nbanana\napple"
+/// ```
+pub fun sort_lines_desc(text: Text): Text {
+    return trust $ sort -r <<< "{text}" $
+}
+
+/// Sorts lines of text numerically.
+///
+/// ### Usage
+/// ```ab
+/// let sorted = sort_lines_numeric("10\n2\n1")
+/// echo(sorted) // "1\n2\n10"
+/// ```
+pub fun sort_lines_numeric(text: Text): Text {
+    return trust $ sort -n <<< "{text}" $
+}
````
Let's combine these functions into one function:
```js
pub fun sort_lines(text: Text, desc: Bool = false, numeric: Bool = false)
```

Ph0enixKM commented on src/std/text.ab:
````diff
@@ -504,3 +504,95 @@ pub fun rpad(text: Text, pad: Text, length: Int): Text {
+    return trust $ sort -n <<< "{text}" $
+}
+
+/// Removes consecutive duplicate lines from text.
+///
+/// ### Usage
+/// ```ab
+/// let result = uniq_lines("foo\nfoo\nbar\nbar\nbaz")
+/// echo(result) // "foo\nbar\nbaz"
+/// ```
+pub fun uniq_lines(text: Text): Text {
+    return trust $ uniq <<< "{text}" $
+}
+
+/// Removes all duplicate lines from text (not just consecutive).
+///
+/// ### Usage
+/// ```ab
+/// let result = uniq_all_lines("foo\nbar\nfoo\nbaz\nbar")
+/// echo(result) // "foo\nbar\nbaz"
+/// ```
+pub fun uniq_all_lines(text: Text): Text {
+    return trust $ awk '!seen[\$0]++' <<< "{text}" $
+}
````
I feel like the following names make more sense to me:
- `uniq_lines_adj`- this only removes adjacent duplicates
- `uniq_lines` - this removes global duplicates

Tirito6626 replied:
i would've suggested making global as boolean argument and then merge both functions into one

Ph0enixKM replied:
@Tirito6626 your idea is even better!

Ph0enixKM reviewed (changes requested):
Great job @rahuld109 ! Thank you for your contribution. This is a very well-written Amber code! Here are some minor suggestions regarding the library functions structure as a whole.

Tirito6626 committed:
Merge branch 'staging' into feat/stdlib-functions

Tirito6626 committed:
Update src/std/text.ab

Co-authored-by: Paweł Karaś <pkaras.it@gmail.com>

Tirito6626 committed:
Update src/std/text.ab

Co-authored-by: Paweł Karaś <pkaras.it@gmail.com>

Tirito6626 committed:
Update src/std/text.ab

Co-authored-by: Paweł Karaś <pkaras.it@gmail.com>

Tirito6626 committed:
Merge branch 'staging' into feat/stdlib-functions

Ph0enixKM commented:
This PR has been completed in https://github.com/amber-lang/amber/pull/965
